src/wallet/rpcwallet.cpp

```diff
@@ -1374,48 +1378,54 @@ UniValue viewshieldedtransaction(const JSONRPCRequest& request)
         if (res == sspkm->mapSaplingNullifiersToNotes.end()) {
             continue;
         }
-        auto op = res->second;
-        auto wtxPrev = pwalletMain->mapWallet.at(op.hash);
-
-        auto decrypted = wtxPrev.DecryptSaplingNote(op).get();
-        auto notePt = decrypted.first;
-        auto pa = decrypted.second;
-
-        // Store the OutgoingViewingKey for recovering outputs
-        libzcash::SaplingExtendedFullViewingKey extfvk;
-        assert(pwalletMain->GetSaplingFullViewingKey(wtxPrev.mapSaplingNoteData.at(op).ivk, extfvk));
-        ovks.insert(extfvk.fvk.ovk);
+        const auto& op = res->second;
+        std::string addrStr = "unknown";
+        UniValue amountStr = UniValue("unknown");
+        CAmount amount = 0;
+        auto wtxPrevIt = pwalletMain->mapWallet.find(op.hash);
+        if (wtxPrevIt != pwalletMain->mapWallet.end()) {
+            const auto ndIt = wtxPrevIt->second.mapSaplingNoteData.find(op);
+            if (ndIt != wtxPrevIt->second.mapSaplingNoteData.end()) {
+                // get cached address and amount
+                if (ndIt->second.address) {
+                    addrStr = KeyIO::EncodePaymentAddress(*(ndIt->second.address));
+                }
+                if (ndIt->second.amount) {
+                    amount = *(ndIt->second.amount);
+                    amountStr = ValueFromAmount(amount);
+                }
+            }
+        }
 
         UniValue entry_(UniValue::VOBJ);
         entry_.pushKV("spend", (int)i);
         entry_.pushKV("txidPrev", op.hash.GetHex());
         entry_.pushKV("outputPrev", (int)op.n);
-        entry_.pushKV("address", KeyIO::EncodePaymentAddress(pa));
-        entry_.pushKV("value", ValueFromAmount(notePt.value()));
-        entry_.pushKV("valueSat", notePt.value());
+        entry_.pushKV("address", addrStr);
+        entry_.pushKV("value", amountStr);
+        entry_.pushKV("valueSat", amount);
         spends.push_back(entry_);
     }
 
     // Sapling outputs
     for (uint32_t i = 0; i < wtx.sapData->vShieldedOutput.size(); ++i) {
         auto op = SaplingOutPoint(hash, i);
-
+        if (!wtx.mapSaplingNoteData.count(op)) continue;
         libzcash::SaplingNotePlaintext notePt;
         libzcash::SaplingPaymentAddress pa;
-        bool isOutgoing;
 
-        auto decrypted = wtx.DecryptSaplingNote(op);
-        if (decrypted) {
+        const bool isOutgoing = !wtx.mapSaplingNoteData.at(op).IsMyNote();
+        if (!isOutgoing) {
+            auto decrypted = wtx.DecryptSaplingNote(op);
+            assert(decrypted);
             notePt = decrypted->first;
             pa = decrypted->second;
-            isOutgoing = false;
         } else {
-            // Try recovering the output
-            auto recovered = wtx.RecoverSaplingNote(op, ovks);
+            // Try recovering with the inputs ovk
+            auto recovered = sspkm->TryToRecoverNote(wtx, op);
```

Comment:
With this change, we don't need to get the ovk in this function at all. Can remove line 1361 to 1366.

Reply:
this is all refactored in a later commit.